### PR TITLE
LADX: Improve icon guesses for foreign items

### PR DESCRIPTION
**Pull request by threeandthreee**
## What is this fixing or adding?
There already exists a system for assigning icons to foreign items based on their name. This fleshes out the list of forbidden words and synonyms to cover a wider range of items and moves the list to a new file (because its long).

## How was this tested?
Packed adjustments into ladx.apworld and playtested.

@zig-for 

As an aside, I'd like to use some other icons such as the piece of power for triforce pieces, but not sure how to do that yet :)

### CODE REVIEW

File: worlds/ladx/ItemIconGuessing.py
Comment (zig-for) [inline worlds/ladx/ItemIconGuessing.py:22]: Why is this forbidden? We want Bombs/Arrows to show up as the proper thing

File: worlds/ladx/ItemIconGuessing.py
Comment (zig-for) [inline worlds/ladx/ItemIconGuessing.py:126]: Ideally the system should be able to pick up on identical named items

**Comment by zig-for** [comment]
🙏 Thank you for this, a few questions. I agree that we should figure out how to get pieces of power working.

File: worlds/ladx/__init__.py
Comment (zig-for) [inline worlds/ladx/__init__.py:386]: Can you change this to "RUPEE"? Currently "1 RUPEE" is turned into a mystery item

File: worlds/ladx/ItemIconGuessing.py
Comment (zig-for) [inline worlds/ladx/ItemIconGuessing.py:63]: Ideally we should check for both "TESTWORD" and "TESTWORDS" in the list, to not have to manually mark up each plural word

File: worlds/ladx/ItemIconGuessing.py
Comment (threeandthreee) [reply worlds/ladx/ItemIconGuessing.py:22]: "BOMBS" hits "MAX_BOMBS_UPGRADE", same for arrows. The BOMB and ARROW icons work better so I associate those instead.

File: worlds/ladx/ItemIconGuessing.py
Comment (threeandthreee) [reply worlds/ladx/ItemIconGuessing.py:126]: Those were in the original so I left them in. I figured it might be to make sure a different type of key doesn't get used instead of the small key. I'll remove it and test.

Comment (threeandthreee) [reply worlds/ladx/ItemIconGuessing.py:126]: P3_ladx_qBWLCehtTk651R1kccoSDQ-230921-145220

This happens with `KEY:KEY` removed. Should be able to safely remove `MAP:MAP` and `COMPASS:COMPASS` though.

Alternatively there's an argument for nightmare keys being the default key icon, and instead doing a specific check for 'SMALL KEY' (like we currently do for 'BIG KEY')

Comment (zig-for) [reply worlds/ladx/ItemIconGuessing.py:126]: Arggh. I knew that was there for a reason. I like the small keys being default, put it back with a comment. My bad for not leaving a comment before.

File: worlds/ladx/ItemIconGuessing.py
Comment (zig-for) [reply worlds/ladx/ItemIconGuessing.py:22]: I guess my worry is making sure that "Power Bombs", "10 Bombs", etc get the right icon. Which BTW we should put a specific case for power bombs if you don't already have it.

Comment (threeandthreee) [reply worlds/ladx/ItemIconGuessing.py:22]: P3_ladx_qBWLCehtTk651R1kccoSDQ-230921-152157

Yeah it works, and power bombs as well with no special rule necessary. (I added `POWER` to the forbidden words list)

Comment (threeandthreee) [reply worlds/ladx/ItemIconGuessing.py:22]: P3_ladx_S0gNPq-_T2GaOA1JCGLmRg-230921-160046

**Comment by threeandthreee** [comment]
P1_ladx_Dc5vfQUQRHq065A29N0Mig-230923-004139

Got pieces of power working, as well as guardian acorns. No animation on the piece of power though, haven't been able to figure that out.

I think it should also be possible to place a bomb arrow, which would be a fun one to use for missiles.

**Comment by zig-for** [comment]
Excellent. Animations on pieces of power probably won't "just work". I'll poke at it, but we might need to flag down daid.

**Comment by zig-for** [comment]
See `chest.asm`, I think the code in use for `renderLargeItem` also needs to be called if you pick up a piece of power.

**Comment by threeandthreee** [comment]
> See `chest.asm`, I think the code in use for `renderLargeItem` also needs to be called if you pick up a piece of power.

The code for color cycling the instruments? I tried that and it didn't seem to do anything. As I understand it, the piece of power should be swapping between two variants rather than color cycling.

I think that this is the piece of power animation code in the disassembly https://github.com/zladx/LADX-Disassembly/blob/main/src/code/entities/bank3.asm#L3024

My attempts to incorporate the code so far have just led to failures to generate, but I'm brand new to assembly.

**Comment by zig-for** [comment]
I can give it a shot later

**Comment by zig-for** [comment]
Bump, is this ok even without anims @threeandthreee

**Comment by threeandthreee** [comment]
Yeah, animation is just an extra as far as I'm concerned.

**Comment by threeandthreee** [comment]
Just to check, is there anything else I can do toward getting this pull request merged?

**Comment by NewSoupVi** [comment]
Zig is no longer actively maintaining LADX.

I'm still willing to merge LADX PRs, but only with a larger volume of peer reviews.

File: worlds/ladx/ItemIconGuessing.py
Comment (ScipioWright) [inline worlds/ladx/ItemIconGuessing.py:202]: Is this todo still valid?

File: worlds/ladx/ItemIconGuessing.py
Comment (ScipioWright) [inline worlds/ladx/ItemIconGuessing.py:189]: 'POWER' appears to be in both the `FORBIDDEN` and `SYNONYMS` dict. Is this intentional? If so, what does this do?

**Review by ScipioWright — COMMENTED** [review]
Would you want more names related to specific games? I can think of quite a few for Tunic

File: worlds/ladx/ItemIconGuessing.py
Comment (threeandthreee) [reply worlds/ladx/ItemIconGuessing.py:202]: Its a comment left in from the original implementation, I'm not entirely sure what zig meant with it tbh.

File: worlds/ladx/ItemIconGuessing.py
Comment (threeandthreee) [reply worlds/ladx/ItemIconGuessing.py:189]: Yes, forbidding the word keeps it from matching 'power bracelet', and then the synonym makes it match the piece of power instead.

**Comment by threeandthreee** [comment]
> Would you want more names related to specific games? I can think of quite a few for Tunic

Yes please

**Comment by ScipioWright** [comment]
For Tunic:
```
'FIRECRACKER': 'BOMB',
'BERRY': 'NUT',  # do HP BERRY and MP BERRY if you want it to only affect the Tunic ones, or take this one out, up to you
'MAGIC ORB': 'HOOKSHOT',
'HP OFFERING': 'TRADING_ITEM_HIBISCUS',  # Tunic: a flower
'HERO RELIC - HP': 'TRADING_ITEM_HIBISCUS',
'MP OFFERING': 'TOADSTOOL',  # Tunic: a mushroom
'HERO RELIC - MP': 'TOADSTOOL',
'SP OFFERING': 'FEATHER',  # Tunic: a feather
'HERO RELIC - SP': 'FEATHER',  
'QUESTAGON': 'PIECE_OF_POWER',  # Tunic: triforce piece equivalent
'INVERTED ASH': 'MEDICINE',  # Tunic: card with a potion on it
"AURA'S GEM": 'SHIELD',  # Tunic: card that enhances the shield
'DUSTY': 'TRADING_ITEM_BROOM',  # Tunic: a broom
'SPRING FALLS': 'TRADING_ITEM_HIBISCUS',  # Tunic: a flower

```
For Noita:
```
'KANTELE': 'HARP',  # Noita: it's a string instrument, vaguely harp-shaped
'ALL-SEEING EYE': 'TRADING_ITEM_MAGNIFYING_GLASS',  # Noita: lets you find secrets
'GOURD': 'TRADING_ITEM_PINEAPPLE',  # Noita
```

**Comment by ScipioWright** [comment]
I'll try to remember to give this a more thorough review and test this weekend o:

File: worlds/ladx/ItemIconGuessing.py
Comment (ScipioWright) [inline worlds/ladx/ItemIconGuessing.py:193]:
```suggestion
    'TRIANGLES': 'PIECE_OF_POWER',
```
Since currently there's a duplicate key with the triangle instrument

Comment (threeandthreee) [reply worlds/ladx/ItemIconGuessing.py:193]: Oh true, i think other cases of "triangle" would probably match piece of power better than the instrument though, so I think I'll just remove the instrument line.

Comment (ScipioWright) [reply worlds/ladx/ItemIconGuessing.py:192]: Can we assume all items can be used as values here? If not, maybe keep that instrument in commented out, with a note that it's unused, so in the future if someone wants to add more they can see that option there

Comment (threeandthreee) [reply worlds/ladx/ItemIconGuessing.py:192]: It is all items but I'll reorganize it for better clarity.

**Review by ScipioWright — APPROVED** [review]
Tested it, it genned fine and runs fine

File: worlds/ladx/ItemIconGuessing.py
Comment (ScipioWright) [inline worlds/ladx/ItemIconGuessing.py:475]:
```suggestion
    'HP OFFERING': 'TRADING_ITEM_HIBISCUS',  # a flower
    'LUCKY UP': 'HEART_CONTAINER',  # card with a heart on it
```
Just another one I thought of o:

File: worlds/ladx/ItemIconGuessing.py
Comment (ScipioWright) [inline worlds/ladx/ItemIconGuessing.py:476]:
```suggestion
    'LUCKY CUP': 'HEART_CONTAINER',  # card with a heart on it
```
Woops

**Comment by zig-for** [comment]
A few notes:
* 1 Rupee was broken before (shown as letter), it's worth checking to see if it's fixed now with better pluralization support
* A few games have different ways of name splitting - IIRC SMZ3 and Hollow Knight use CamelCase, worth special casing somehow

**Comment by threeandthreee** [comment]
rupee

added rupee synonym and improved word splitting

**Comment by threeandthreee** [comment]
This could have been a separate PR but I wanted to use the Guardian Acorn and Piece of Power.

Added a yaml option to simply indicate whether or not foreign items are progression, where progression items are a Piece of Power and other items are Guardian Acorns. Went with these two since they don't apply to any local items.

Was mentioned on discord and is pretty simple to put in.
https://discord.com/channels/731205301247803413/1282851721420210318/1282856567351541790

image
image
image

**Comment by NewSoupVi** [comment]
For Witness, there are three major item groups

Every item in the "Doors" item group should resolve to a Key
Except for "Boat", which should be, like the angler hook or flippers or sth idk

Every item in the "Lasers" item group could resolve to... well, a random instrument, tbh? Since that's the equivalent, it's the "main McGuffin".

Off of vibes, without thinking about it too much (feel free to suggest sth different):
Symmetry Laser - Triangle
Desert Laser - Marimba
Quarry Laser - Drums
Shadows Laser - Violin
Keep Laser - Organ
Town Laser - Bell
Monastery Laser - Violin
Jungle Laser - Harp
Bunker Laser - Harp (idk, it's a greenhouse full of flowers)
Swamp Laser - Horn
Treehouse Laser - Horn

Every item in the Symbols item group could probably just be the map. Although some of them could have better associations, it's just hard to find one for all of them

**Comment by NewSoupVi** [comment]
Ftr, I believe the door group would be covered by the following phrases:

"Door"
"(Door)"
"Panel"
"(Panel)"
"Doors"
"Panels"
"Water Pumps"
"Gates"

**Comment by threeandthreee** [comment]
It already pluralizes (well, sticks 's' at the end), so 'DOOR' will work for both 'DOOR' and 'DOORS'. Off the top of my head I can't think of what it does with parens, but would make sense to filter those out too.

File: worlds/ladx/ItemIconGuessing.py
Comment (ScipioWright) [inline worlds/ladx/ItemIconGuessing.py:70]: Would recommend against this one. Lots of KH2 items have "Boost" in the name, like Combo Boost or "Item Boost" which have nothing to do with boots

File: worlds/ladx/ItemIconGuessing.py
Comment (ScipioWright) [inline worlds/ladx/ItemIconGuessing.py:100]: This one has way fewer uses, though it's still a bit questionable -- Fire Surge and Acid Surge spells in DS3, Chrono Surge in SC2

File: worlds/ladx/ItemIconGuessing.py
Comment (NewSoupVi) [inline worlds/ladx/ItemIconGuessing.py:424]: Witness stuff looks good!!!

```suggestion
    'TREEHOUSE LASER': 'INSTRUMENT2',
    'WATER PUMPS': 'KEY',
```

"Swamp Water Pumps" is the only set of doors that does not have "Door" in the name, but it still acts the same as all the other ones, so this would make sense

Comment (NewSoupVi) [reply worlds/ladx/ItemIconGuessing.py:424]: Oh nvm actually there is also 
```suggestion
    'TREEHOUSE LASER': 'INSTRUMENT2',
    'CAVES SHORTCUTS': 'KEY',
    'KEEP SHORTCUTS': 'KEY',
    'MONASTERY SHORTCUTS': 'KEY',
    'SWAMP SHORTCUTS': 'KEY',
```

Unless you think "Shortcuts" is general enough of a word to just add this as a generic

**Comment by NewSoupVi** [comment]
I realized there's also a couple Door items called "Shortcuts"

https://github.com/ArchipelagoMW/Archipelago/pull/2201#discussion_r1769170355

But after that I think it's truly all, I had a quick glance over the full list

**Review by SushiKishi — APPROVED** [review]
Seeds with this implemented have generated as expected.

I have personally only used the progression/non-progression setting with this code implemented, so I can't comment on the "name => icon" selection side of this change. 

I was going to suggest that the PoP icon should blink, as the 'vanilla' one does, but it leads back to confusion -- is that a Piece of Power, or is it an AP item drop? Having it non-blinking makes it obvious which is which.

I'd rather just get, I dunno, a star like in LTTP, but I realize options are limited within the confines of a GBC cart, and if there were no further additions to the feature I'd still be using the Progression/Non-Progression option 100% of the time.

**Comment by threeandthreee** [comment]
In vanilla LADX, when you find a small key that drops from the ceiling, you get no text box when you pick it up. The randomizer has kept this behavior but its a problem when we're customizing foreign items because if it looks like a small key, its treated as a small key, so the text box can get skipped when it shouldn't be.

Adding a change to this branch that removes the handling to skip text on small key drops. We polled the channel and the majority is in favor of this. 

https://discord.com/channels/731205301247803413/1090819435893362768/1300544655435370607

**Comment by threeandthreee** [comment]
> I was going to suggest that the PoP icon should blink, as the 'vanilla' one does, but it leads back to confusion -- is that a Piece of Power, or is it an AP item drop? Having it non-blinking makes it obvious which is which.

An attempt was made earlier, I couldnt figure out the animation :smile:

> I'd rather just get, I dunno, a star like in LTTP, but I realize options are limited within the confines of a GBC cart, and if there were no further additions to the feature I'd still be using the Progression/Non-Progression option 100% of the time.

A custom icon would certainly be possible but its not something I'm personally looking to work on

**Review by palex00 — APPROVED** [review]
This PR was tested as part of a Beta apworld that was used in three dedicated Test Asyncs, as well as dozen of normal Syncs. I have also personally run 200 test generations which succeeded.

We have also done polling on what users want, etc. which [can be found here](https://docs.google.com/forms/d/e/1FAIpQLSdLJ_j9RiS5TRjtovV_O-HwPEJMygLTAFotw3AdlUxfMCByPQ/viewanalytics)

The new feature has worked as expected and other stuff does not seem impacted.

File: worlds/ladx/Options.py
Comment (NewSoupVi) [inline worlds/ladx/Options.py:513]: Minor thing, but in my opinion, options other than Toggle and DefaultOnToggle should generally explicitly include their default

*Low priority*

File: worlds/ladx/__init__.py
Comment (NewSoupVi) [inline worlds/ladx/__init__.py:394]: First of all, you could do this:

```suggestion
            pluralizations = {}
            for k, v in self.name_cache.items():
                ks = k + 'S'
                if ks not in self.name_cache:
                    pluralizations[ks] = v
            self.name_cache.update(pluralizations)

```

Comment (NewSoupVi) [reply worlds/ladx/__init__.py:394]: But also, here's a cute idea for how to make this code shorter:

```suggestion
            pluralizations = {k + "S": v for k, v in self.name_cache.items()}
            self.name_cache = pluralizations | self.name_cache
```

The reason this works is that by putting `self.name_cache` on the right side of the `|`, you're preferring its values over the ones from `pluralizations`

Anyway. Not too important. I would prefer if you gave your variables longer names, though, generally.

*Low priority*

File: worlds/ladx/__init__.py
Comment (NewSoupVi) [inline worlds/ladx/__init__.py:404]:
```suggestion
        for phrase, icon in phrases.items():
            if phrase in uppered:
                return icon
```

Again, not super important.
Just trying to get you comfortable with the `.items()` iterator for dictionaries.

*Low priority*

**Review by NewSoupVi — APPROVED** [review]
Seems good. Did a bunch of generations together with the async filler yaml.

I have a *couple* of minor comments about code quality, but they're not that important - Mostly putting these here because I know threeandthree is an aspiring world maintainer so I want to give them some learning opportunities

Anyway, I'll probably just merge this soon:tm:, unless there are any objections. More of these associations can be added in future PRs and they should be much simpler.

**Comment by NewSoupVi** [comment]
I forgor lol

**Pull request merged**
